File: Dynamic programming/Range Sum Query - Immutable.py
from typing import List


class NumArray:

    def __init__(self, nums: List[int]):
        sumNum = 0
        for pos in range(0, len(nums)):
            sumNum += nums[pos]
            nums[pos] = sumNum
        self.arr = nums[:]

    def sumRange(self, i: int, j: int) -> int:
        if i == 0:
            return self.arr[j]
        else:
            return self.arr[j] - self.arr[i-1]

print("Enter val of list")
arr = list(map(int, input().rstrip().split()))
print("Enter i and j")
start = int(input())
end = int(input())
obj = NumArray(arr)
sumForRange = obj.sumRange(start, end)
print(sumForRange)

File: August 30 day challenge '21/RangeSumQueryImmutable.java
import java.io.*;
import java.util.*;

class NumArray {
    int[] arr;

    public NumArray(int[] nums) {
        arr = nums.clone();
        for (int i = 1; i < nums.length; i++)
            arr[i] += arr[i - 1];
    }

    public int sumRange(int left, int right) {
        int sum = arr[right];
        sum -= left == 0 ? 0 : arr[left - 1];
        return sum;
    }
}
